### PR TITLE
ip: Add support for locating Mibbit users

### PR DESCRIPTION
**Pull request by kwaaak**
Depending on the relationship between Mibbit and the IRC network, the IP address will either be displayed normally in the hostmask or represented in hexadecimal format in the user field.
This pull request adds logic that detects and converts this IP address.

### CODE REVIEW

File: sopel/modules/ip.py [hunk 1]
```diff
@@ -127,9 +128,16 @@ def ip(bot, trigger):
...
+            if user_in_botdb.hostmask.endswith(".mibbit.com"):
+                hexip = user_in_botdb.hostmask.split("!")[1].split("@")[0]
+                # Stolen from https://stackoverflow.com/a/2198052, converts hexip to dotted quad
+                addr_long = int(hexip, 16)
+                query = socket.inet_ntoa(struct.pack(">L", addr_long))
+            else:
+                query = user_in_botdb.host
         else:
-            return bot.say("I am not aware of this user.")
+            return bot.say("I\'m not aware of this user.")
```
Comment (dgw) [inline sopel/modules/ip.py]: The `\` is unnecessary for single quotes inside double-quoted strings.

This particular change is also unrelated to adding Mibbit hexip support, and "I\\'m" not sure it's worth changing to a contraction anyway. "I am not aware" sounds fine?

File: sopel/modules/ip.py [hunk 2]
```diff
@@ -127,9 +128,16 @@ def ip(bot, trigger):
         username = trigger.group(2).strip()
         user_in_botdb = bot.users.get(username)
         if user_in_botdb is not None:
-            query = user_in_botdb.host
+            # Mibbit on not webirc-enabled networks https://wiki.mibbit.com/index.php/Hexip#What_is_shown_when_connected_to_an_IRC_Network_that_is_not_Enabled
```
Comment (dgw) [inline sopel/modules/ip.py]: Make this comment two lines, breaking before the URL.

File: sopel/modules/ip.py [hunk 3]
```diff
@@ -127,9 +128,16 @@ def ip(bot, trigger):
...
+            if user_in_botdb.hostmask.endswith(".mibbit.com"):
+                hexip = user_in_botdb.hostmask.split("!")[1].split("@")[0]
+                # Stolen from https://stackoverflow.com/a/2198052, converts hexip to dotted quad
+                addr_long = int(hexip, 16)
+                query = socket.inet_ntoa(struct.pack(">L", addr_long))
```
Comment (dgw) [inline sopel/modules/ip.py]: What does this do for IPv6 addresses? Hex-encoded IPv6 would be too long for the `user` part of the hostmask (max ident length seems to always be 10 characters, and anything past that is simply truncated), so this will only work on IPv4 addresses even if Mibbit tries to send hexip for v6.

Since an IPv4 hexip address should always be 8 characters, I'd say this branch should bail if it's the wrong length. Maybe stick the `query = user_in_botdb.host` assignment before the `if` and drop the `else`?

File: sopel/modules/ip.py (hunk 1 above)
Comment (kwaaak) [reply sopel/modules/ip.py]: Not on purpose, my local copy must be different

File: sopel/modules/ip.py (hunk 3 above)
Comment (kwaaak) [reply sopel/modules/ip.py]: mibbit.com and chat.mibbit.com are not reachable via IPv6

Comment (dgw) [reply sopel/modules/ip.py]: Other webirc/CGI:IRC clients might be. See the top-level review comment, which asks why only Mibbit should be supported for hexip lookups.

**Comment by kwaaak** [comment]
Hopefully I can make this more universal soon.
I am having problems finding CGI:IRC instances that are available via IPv6. So far I have only found https://www.oftc.net/ that might support it.

**Comment by dgw** [comment]
I spent about half an hour trying to get an IPv6 tunnel running on my local system with no luck, and had to give it up for tonight. Can neither confirm nor deny that OFTC's webchat actually works over IPv6, but it does resolve according to `dig webchat.oftc.net aaaa`…
